source/docs/design/rw2022-talk-proposal.tex

```diff
@@ -0,0 +1,199 @@
+
+We consider the Rust's type system and borrow checker a central
+component in Verus' design and ability to generate verficiation
+conditions that Z3 can solve efficiently. Linear types mean that we can
+often encode potentially mutable, complex objects, as immutable SMT
+datatypes, rather than having to rely on complex rules and axioms. Even
+objects with interior mutability that expose an ``immutable'',
+deterministic interface can be often safely encoded as Z3 values. The
+borrow checker ensures that each value is either (i) currently uniquely
+owned, (ii) borrowed many times but immutably, or (iii) borrowed mutably
+exactly once.
+
+\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}
+
+Example. A simple Rust function with a functional spec and the generated
+Z3 verification condition.
+
+Rust:
+
+\begin{verbatim}
+fn small_sum(a: u64, b: u64) -> u64 {
+    requires([
+        a < 128,
+        b < 128,
+    ]);
+    ensures(|result: u64| result < 256);
+    a + b
+}
+\end{verbatim}
+
+Z3:
+
+\begin{verbatim}
+  (declare-const result@ Int)
+  (declare-const a@ Int)
+  (declare-const b@ Int)
+  ... (3 lines with heuristics machinery and u64 type invariants) ...
+  (assert (< a@ 128))
+  (assert (< b@ 128))
+  (assert
+   (not (=>
+     (= result@ (uClip 64 (+ a@ b@))) (< result@ 256)
+  )))
+\end{verbatim}
+
+TODO, maybe replace with something with an ADT, but in the same spirit?
+TODO: get rid of the example in the proposal.
+
+\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}
+
+From our experience, Rust's memory management story, based on linear
+types that prevents mutable aliased references, removes a significant
+challenge of other tools, like Dafny, that need to encode and manage the
+heap. Heap-reasoning is notoriously costly in SMT-based tools. In our
+experience building and verifying VeriBetrKV, a verified crash-safe
+high-performance key-value store, we were able to reduce proof burden
+(in terms of LOCs and verification time) by up to 30\% by integrating a
+linear type system into a fork of Dafny. This advantage comes naturally
+in a language designed around linearity, like Rust.
+
+\section{Borrow-checking proofs}\label{borrow-checking-proofs}
+
+As part of recent work that's under submission, we recognized the value
+of tracking ownership of ``ghost'' variables, that do not appear in the
+emitted code, but aid verification. For example, in concurrent code,
+linear ``ghost'' variables permit tracking of ``ghost'' permissions to
+access non-atomic memory, they can represent a thread's knowledge of a
+concurrent, shared-memory protocol state, or they can represent limited
+aliasing domains, for example as a mechanism to encapsulate and prove
+non-linear data-structures that require aliased references such as
+\texttt{Rc}. To enable using linearity to track these separation logic
+assertions Verus supports three modes for variables and functions, as
+shown in Example X.
```

Comment:
this paragraph is similar in spirt to the one before. maybe you can consolidate or at least move them to the same section